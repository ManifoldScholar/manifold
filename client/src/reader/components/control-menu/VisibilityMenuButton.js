import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

export default class VisibilityMenuButton extends PureComponent {
  static displayName = "ControlMenu.VisibilityMenuButton";

  static propTypes = {
    toggle: PropTypes.func,
    active: PropTypes.bool
  };

  clickHandler = event => {
    event.stopPropagation();
    this.props.toggle();
  };

  render() {
    const buttonClass = classNames({
      "button-visibility": true,
      "button-active": this.props.active
    });

    return (
      <button
        className={buttonClass}
        onClick={this.clickHandler}
        data-id="toggle-visibility"
      >
        <i className="manicon manicon-eye-outline" aria-hidden="true" />
        <span className="screen-reader-text">Open the visibility menu</span>
      </button>
    );
  }
}
