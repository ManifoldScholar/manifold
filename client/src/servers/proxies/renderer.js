import config from "config";
import proxy from "http-proxy-middleware";
import ch from "../../helpers/consoleHelpers";
import createStore from "../../store/createStore";
import exceptionRenderer from "../../helpers/exceptionRenderer";
import manifoldBootstrap from "bootstrap";
import CookieHelper from "helpers/cookie/Server";

const ssrRenderUrl = `http://${config.services.client.domain}:${config.services.client.sparePort}`;

export default function makeRendererProxy(requestHandler) {
  const reactServerProxy = proxy({
    target: ssrRenderUrl,
    changeOrigin: true,
    logLevel: "silent",

    onError: (err, req, res) => {
      ch.error(`Error proxying to port ${ssrRenderUrl}`);
      ch.error(`Perhaps the SSR service is in the process of reloading?`);
      ch.error(`Falling back to SSR rescue service only`);

      const store = createStore();

      const render = () => {
        try {
          requestHandler(req, res);
        } catch (error) {
          if (error.code === "MODULE_NOT_FOUND") {
            const msg =
              "Waiting for initial Webpack build to complete. Wait a few seconds " +
              "and reload.";
            ch.error(msg);
            res.setHeader("Content-Type", "text/html");
            res.end(msg);
          } else {
            ch.error(
              `Rendering fallback failed to render inn server-developmet.js`
            );
            res.setHeader("Content-Type", "text/html");
            res.end(exceptionRenderer(error));
          }
        }
      };

      const promises = [];
      const cookie = new CookieHelper(req, res);
      promises.push(manifoldBootstrap(store.getState, store.dispatch, cookie));
      Promise.all(promises).then(render, render);
    }
  });
  return reactServerProxy;
}
