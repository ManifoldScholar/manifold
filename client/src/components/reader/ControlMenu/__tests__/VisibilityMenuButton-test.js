import React from "react";
import renderer from "react-test-renderer";
import VisibilityMenuButton from "../VisibilityMenuButton";
import Adapter from "enzyme-adapter-react-16";
import Enzyme from "enzyme";

Enzyme.configure({ adapter: new Adapter() });

describe("Reader.ControlMenu.VisibilityMenuButton Component", () => {
  const toggleMock = jest.fn();

  const root = <VisibilityMenuButton active={false} toggle={toggleMock} />;

  it("renders correctly", () => {
    const component = renderer.create(root);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should trigger toggle callback when toggle is clicked", () => {
    const wrapper = Enzyme.mount(root);
    toggleMock.mockClear();
    wrapper.find('[data-id="toggle-visibility"]').first().simulate("click");
    expect(toggleMock).toHaveBeenCalled();
  });
});
