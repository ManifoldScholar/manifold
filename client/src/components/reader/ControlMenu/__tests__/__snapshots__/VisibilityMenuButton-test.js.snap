// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reader.ControlMenu.VisibilityMenuButton Component renders correctly 1`] = `
<button
  className="button-visibility"
  data-id="toggle-visibility"
  onClick={[Function]}
>
  <i
    aria-hidden="true"
    className="manicon manicon-eye-outline"
  />
  <span
    className="screen-reader-text"
  >
    Open the visibility menu
  </span>
</button>
`;
